### PR TITLE
Ask fewer questions during `init`.

### PR DESCRIPTION
**Pull request by ellismg**
This change reworks environment initialization to no longer require a subscription or location.  Instead, these properties are now state that are managed by the individual providers and stored in the environment (long term, we envision other provisioning providers which don't need a subscription id or location, which is why each provider manages this state instead of the `provisoning.Manager` type).

Since `init` and `new` no longer configure either the subscription or location, they no longer require you to be logged into Azure.

In addition, `AZURE_PRINCIPAL_ID` is no longer added to the `.env` file.  Instead, the value is present when evaluating the parameters template file (this means that if you log in as a different principal from the one you created the environment with, when you run `azd provision`, you'll get the id of the current principal, not the old principal, which makes more sense) but not persisted to disk.

Components which need to call either `GetSubscriptionId` or `GetLocation` must now guard against the case where these values are not set, because `provision` (or `env refresh`) have not been run yet. In these cases, we tell the user the command can not run because infrastructure has not been provisioned and guide them to run `azd provision`.

### CODE REVIEW

**Comment by vhvb1989** [comment]
> For existing commands like deploy and pipeline config which previously depended on these values

Well, I get it for `deploy`. But it is a shame for `pipeline config`.  Can we make `pipeline config` to also ask for sub-id and location if missing?  
I don't think it is a good experience to force users to run `provision` locally.

**Comment by ellismg** [comment]
> Can we make pipeline config to also ask for sub-id and location if missing?
> I don't think it is a good experience to force users to run provision locally.

I continue to think about this - but it's something we'll have to relax in the future.  I think the code is structured in a way such that it would be fairly straightforward to allow.

**Comment by jongio** [comment]
A common scenario is sharing envs. So a developer shouldn't be required to run provision in order to deploy. 

Steps 
Azd env refresh 
Azd deploy

(No provision on that machine)

**Comment by ellismg** [comment]
> Azd env refresh
> Azd deploy

This continues to work.  `azd env refresh` will prompt for a subscription and location and then refresh as it would have done before and `deploy` will work then.

**Review by weikanglim — APPROVED** [review]
Overall looks good to me. I love the changes, and also the small code hygiene improvements made along the way. None of those small changes detracted from the overall review.

I'll reflect upon my own honoring of "boy scout rules" -- I've been hesitant with avoiding introducing any additional changes as part of a feature change, but I can see now that it's really possible to make small subtle changes that are overall readable (and with test coverage).

**Review by vhvb1989 — APPROVED** [review]

File: cli/azd/pkg/infra/provisioning/util.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+// for values if they are not.
+func EnsureSubscriptionAndLocation(ctx context.Context, env *environment.Environment, prompters Prompters) error {
+	if env.GetSubscriptionId() == "" {
+		subscriptionId, err := prompters.Subscription(ctx, "Please select an Azure Subscription to use:")
+		if err != nil {
+			return err
+		}
+
+		env.SetSubscriptionId(subscriptionId)
+		telemetry.SetGlobalAttributes(fields.SubscriptionIdKey.String(env.GetSubscriptionId()))
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/pkg/infra/provisioning/util.go:27]: Thanks!  I did not notice that we had code in the `env` that was doing this.  I agree it's a better place - I'll move it.

File: cli/azd/cmd/deploy.go [hunk 2]
```diff
@@ -176,6 +176,12 @@ func (d *deployAction) Run(ctx context.Context) (*actions.ActionResult, error) {
 		return nil, fmt.Errorf("service name '%s' doesn't exist", targetServiceName)
 	}
 
+	if d.env.GetSubscriptionId() == "" {
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/cmd/deploy.go:179]: Yeah - would be good to make some improvements here - opened https://github.com/Azure/azure-dev/issues/1829 to track.

File: cli/azd/pkg/infra/provisioning/util.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/pkg/infra/provisioning/util.go:27]: Centralized everything such that we only call it here now.

File: cli/azd/cmd/util.go [hunk 3]
```diff
@@ -124,40 +121,37 @@ func loadEnvironmentIfAvailable() (*environment.Environment, error) {
 
 func loadOrInitEnvironment(
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/cmd/util.go:122]: These names are much better!  Thank you!

File: cli/azd/cmd/util.go [hunk 4]
```diff
@@ -167,193 +161,50 @@ func loadOrInitEnvironment(
...
 
-	return nil
-}
-
-func getSubscriptionOptions(ctx context.Context, subscriptions account.Manager) ([]string, any, error) {
-	subscriptionInfos, err := subscriptions.GetSubscriptions(ctx)
-	if err != nil {
-		return nil, nil, fmt.Errorf("listing accounts: %w", err)
+	if env.GetSubscriptionId() != "" {
+		telemetry.SetGlobalAttributes(fields.SubscriptionIdKey.String(env.GetSubscriptionId()))
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/cmd/util.go:204]: Indeed - it does!

File: cli/azd/cmd/util.go [hunk 5]
```diff
@@ -167,193 +161,50 @@ func loadOrInitEnvironment(
...
-		accountManager,
-		userProfileService,
-		subResolver); err != nil {
-		return nil, fmt.Errorf("initializing environment: %w", err)
-	}
-
 	if isNew {
-		if err := azdCtx.SetDefaultEnvironmentName(*environmentName); err != nil {
-			return nil, fmt.Errorf("saving default environment name: %w", err)
+		if env.GetEnvName() == "" {
```
[the remark this answers is not in the document]
Comment (ellismg) [reply cli/azd/cmd/util.go:190]: I will do a follow up PR to try to clean this method up even more - I am looking at it and think there's some more simplification that can be done, but I want to give myself some breathing room before jumping in.

**Comment by anfibiacreativa** [comment]
As discussed with @Petermarcu  and @rajeshkamal5050 I would like to revive this discussion, specifically in connection with the `init` command. Right now, when initializing a template, developers are required to provide an environment name right away. This is problematic when considering the possibility of initializing the template in environments like codespaces. From a DX perspective, I feel this decision should only be taken much later in the flow, when there is a clear intention to provision and deploy. Happy to continue the conversation.

**Comment by vhvb1989** [comment]
@anfibiacreativa , can you elaborate on this:

> This is problematic when considering the possibility of initializing the template in environments like codespaces

Usually, when you create/open a codespace from a repo that contains an azure.yaml (azd project), you don't need to run `azd init`. You would just need
- azd auth login (if not done before)
- azd provision

You want to run `azd init` when you either:
- You want azd to pull a project from a git repo to your current folder (usually an empty folder). So you run `azd init -t git-repo-url`
- You are a template-author and want to start an empty azd project.  You run `azd init` an starter template.
- You want to azdfy your existing project, so you run `azd init` and tell azd to auto-detect your existing code and self-configure. This is the path known as `easy init` which is common for .NET Aspire projects.

But, if you are the end user and you are picking codespaces to try/use one azd project (sample or template). You don't need to run azd init (unless the sample/template depends on `easy init` on purpose, like some .NET Aspire projects).   

If the codespace you created and opened already contains an azure.yaml, the only thing that `azd init` will do is to check if there is already an azd environment and if not, ask a name to create it.
If you don't run azd init and run `azd provision` directly, azd will start by asking for an environment name.

**Comment by anfibiacreativa** [comment]
> @anfibiacreativa , can you elaborate on this:
> 
> > This is problematic when considering the possibility of initializing the template in environments like codespaces
> 
> Usually, when you create/open a codespace from a repo that contains an azure.yaml (azd project), you don't need to run `azd init`. You would just need
> 
> * azd auth login (if not done before)
> * azd provision
> 
> You want to run `azd init` when you either:
> 
> * You want azd to pull a project from a git repo to your current folder (usually an empty folder). So you run `azd init -t git-repo-url`
> * You are a template-author and want to start an empty azd project.  You run `azd init` an starter template.
> * You want to azdfy your existing project, so you run `azd init` and tell azd to auto-detect your existing code and self-configure. This is the path known as `easy init` which is common for .NET Aspire projects.
> 
> But, if you are the end user and you are picking codespaces to try/use one azd project (sample or template). You don't need to run azd init (unless the sample/template depends on `easy init` on purpose, like some .NET Aspire projects).
> 
> If the codespace you created and opened already contains an azure.yaml, the only thing that `azd init` will do is to check if there is already an azd environment and if not, ask a name to create it. If you don't run azd init and run `azd provision` directly, azd will start by asking for an environment name.

Thanks for the detailed explanation, but the bullet points do not correspond with the scenario in mind. In the scenario we're discussing, we're opening a Codespace to initialize a template from it. This is in design, right now. I can approach you once the spec is documented. For that particular scenario, that doesn't exist today, `azd init` passing the template flag, will run from a post-create command.

**Comment by vhvb1989** [comment]
@anfibiacreativa , can you try using no prompt and environment name when using azd from a devcontainer hook?
something like: `azd init -t <template-url> --no-prompt -e <suggestedEnvName>`
I tried this: `azd init -t todo-nodejs-mongo --no-prompt -e dev` and azd skip all prompts and creates an initial env named dev.

Can you use it like that?

**Pull request merged**